### DISCUSSION TITLE
Btree Memory Errors

### DISCUSSION THREAD

**Discussion by ajacobs-github** [micropython/micropython#14132] [Libraries & Drivers]
I am encountering intermittent [Errno 12] ENOMEM errors when assigning a key-value in Btree.  It occurs when testing in a loop just after importing several custom python classes with gc.collect() after each import, and before creating any objects.  The free memory on the heap is >81K before assigning the key value. They key is 64B and the value is 512B.  Using OpenMV micropython on STM32 H7.

Before DB Write Memory free heap: 81216 B
Success DB Write
Before DB Write Memory free heap: 81216 B
ERROR DB Write [Errno 12] ENOMEM
Before DB Write Memory free heap: 81216 B
ERROR DB Write [Errno 12] ENOMEM
Before DB Write Memory free heap: 81216 B
Success DB Write
Before DB Write Memory free heap: 81216 B
Success DB Write

This is my funciton to write test key-values to BTree.

```py
def writeBtree(filePathDb, value, key):
    success = False
    key = randstr(64)
    value = randstr(512)
    print("DB Write: Key", key, ", Value", value)
    try:
        dbFile = open(filePathDb, "r+b")
    except:
        dbFile = open(filePathDb, "w+b")
    db = btree.open(dbFile, cachesize = 1024)
    gc.collect()
    print("Before DB Write Memory free heap: %d B" % gc.mem_free())
    try:
        db[key] = value
    except Exception as error:
        print("ERROR DB Write ", error)
        success = False
    else: 
        print("Success DB Write ")
        db.flush()
        db.close()
        dbFile.close()
        success = True
    return success

```

(Edited by @jimmo to add formatting)

**Comment by jimmo** [comment]
@ajacobs-github Please see [Welcome (Please read before posting)](https://github.com/orgs/micropython/discussions/9111) for how to format code for GitHub. (Use triple-backticks)

**Comment by jimmo** [comment]
@ajacobs-github I can't replicate this on ESP32 or Unix (although I might be using your test function incorrectly), and we don't support btree on stm32.

Can you please provide more detail about how to replicate this?
